### PR TITLE
[KERNEL32] Implement GetFinalPathNameByHandleW

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

This PR implements GetFinalPathNameByHandleW. The wine version is broken and doesn't work, because it relies on a broken behavior of NtQueryObject.

## Proposed changes

- [PSDK] Fix up lmshare.h
- [NTDLL_APITEST] Add another test for NtQueryObject (This shows that wine's code doesn't work)
- [KERNEL32_APITEST] Add tests for GetFinalPathNameByHandleW
- [KERNEL32_VISTA] Implement GetFinalPathNameByHandleW

## Testbot runs (Filled in by Devs)

- ✅ KVM x86: https://reactos.org/testman/compare.php?ids=102622,102628,102631
- ✅ KVM x64: https://reactos.org/testman/compare.php?ids=102626,102629,102632

### CODE REVIEW

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    if (!NT_SUCCESS(RtlAppendUnicodeToString(DosVolumeName, L"\\\\?\\")) ||
+        !NT_SUCCESS(RtlAppendUnicodeToString(DosVolumeName, VolumePaths->MultiSz) ||
+        !NT_SUCCESS(RtlAppendUnicodeToString(DosVolumeName, L"\\"))))
+    {
+        DPRINT1("RtlAppendUnicodeToString failed: %lx\n", Status);
+        goto Exit;
+    }
+
+    /* This is based on Windows behavior */
+    SetLastError(0);
```
Comment (oleg-dubinskiy) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]:
```suggestion
    SetLastError(ERROR_SUCCESS);
```
From other side, is setting this one needed at all, if `GetFinalPathNameByHandleW()`, which only uses this function, already sets an appropriate error code below in both success and failure cases? All last error stuff is already managed by the implementation below.
Can it be used by something other too later? Otherwise, if not, then need we to have set it twice?

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+        }
+
+        RtlCopyMemory(lpszFilePath, VolumeBuffer, VolumeLength * sizeof(WCHAR));
+        RtlCopyMemory(lpszFilePath + VolumeLength,
+                      RelativePath.Buffer,
+                      RelativePath.Length);
+        lpszFilePath[FinalLength] = UNICODE_NULL;
+
+        /* This is based on Windows behavior */
+        SetLastError(0);
```
Comment (oleg-dubinskiy) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: Using the proper error code looks much more human-readable than just zero.
```suggestion
        SetLastError(ERROR_SUCCESS);
```

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    WCHAR Buffer[MAX_PATH];
+    PMOUNTMGR_TARGET_NAME TargetName = (PMOUNTMGR_TARGET_NAME)Buffer;
+    PMOUNTMGR_VOLUME_PATHS VolumePaths = (PMOUNTMGR_VOLUME_PATHS)Buffer;
+    NTSTATUS Status;
+
+    /* Validate parameters */
+    if ((DeviceName == NULL) || (DeviceName->Buffer == NULL) || (DeviceName->Length == 0) ||
+        (DosVolumeName == NULL) || (DosVolumeName->Buffer == NULL) || (DosVolumeName->MaximumLength == 0))
+    {
+        ASSERT(FALSE);
```
Comment (oleg-dubinskiy) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c:38]: Maybe it's not necessarily to have this assert? :thinking: Some invalid behaviour testcases (with invalid parameters passed) may get here and crash (in case they are exist). So I'd rather add a `DPRINT1()` with error here instead.

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("QueryDosVolumeNameForNtDeviceName failed: %lx\n", Status);
+        BaseSetLastNTError(Status);
+        return 0;
+    }
+
+    if (VolumeType == VOLUME_NAME_DOS)
+    {
+        /* Calculate the final length and vaidate the buffer size */
```
Comment (learn-more) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]:
```suggestion
        /* Calculate the final length and validate the buffer size */
```
(nitpicking)

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 2 above)
Comment (learn-more) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: Since this is the only case where this appears to be done, I'd assume it comes from `GetVolumeNameForVolumeMountPointW`

**Review by learn-more — APPROVED** [review]
Very nice testcases

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+    /* For everything else we need to split the full path */
+    PathSplit = wcschr(&NameInfo->Name.Buffer[sizeof("\\Device\\") - 1], L'\\');
+    if (PathSplit == NULL)
+    {
+        DPRINT1("Invalid object name: %wZ\n", &NameInfo->Name);
+        SetLastError(ERROR_INVALID_HANDLE);
+        return 0;
+    }
+
+    *PathSplit = UNICODE_NULL;
```
Comment (oleg-dubinskiy) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: This looks a bit confusing: first you're storing some character in this string pointer, and then, on success, setting it to a null-terminator later. Then you're goring on one character forward at line 200 by incrementing it. But since there is a null-terminator, nothing follows after that, so it won't work properly. Moreover, comment on the same line states that there should be a backslash.
So you might wan to initialize that above the `wcschr()` call instead, as follows:
```suggestion
    *PathSplit = UNICODE_NULL;

    PathSplit = wcschr(&NameInfo->Name.Buffer[sizeof("\\Device\\") - 1], L'\\');
    if (PathSplit == NULL)
    {
        DPRINT1("Invalid object name: %wZ\n", &NameInfo->Name);
        SetLastError(ERROR_INVALID_HANDLE);
        return 0;
    }
```
Fix me if I'm wrong somewhere.

Comment (learn-more) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: You can't write to `*PathSplit` before initializing `PathSplit`

Comment (oleg-dubinskiy) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: OK, my mistake then. Sorry.

File: modules/rostests/apitests/ntdll/NtQueryObject.c [hunk 6]
```diff
@@ -99,9 +99,42 @@ START_TEST(NtQueryObject)
...
+
+    /* Query the name of the file */
+    Status = NtQueryObject(FileHandle,
+                           ObjectNameInformation,
+                           &ObjectNameBuffer,
+                           sizeof(ObjectNameBuffer),
+                           &BufferSize3);
+    ok_ntstatus(Status, STATUS_SUCCESS);
+
+    /* Validte that the name starts with "\\Device" */
```
Comment (HBelusca) [inline modules/rostests/apitests/ntdll/NtQueryObject.c]:
```suggestion
    /* Validate that the name starts with "\\Device" */
```

File: modules/rostests/apitests/ntdll/NtQueryObject.c [hunk 7]
```diff
@@ -99,9 +99,42 @@ START_TEST(NtQueryObject)
...
+    ok(wcsncmp(ObjectName->Buffer, L"\\Device", 7) == 0,
+        "ObjectName->Buffer: '%S' does not start with '\\Device'\n",
+        ObjectName->Buffer);
```
Comment (HBelusca) [inline modules/rostests/apitests/ntdll/NtQueryObject.c]:
```suggestion
    ok(wcsncmp(ObjectName->Buffer, L"\\Device", 7) == 0,
       "ObjectName->Buffer: '%S' does not start with '\\Device'\n",
       ObjectName->Buffer);
```

File: modules/rostests/apitests/ntdll/NtQueryObject.c [hunk 8]
```diff
@@ -99,9 +99,42 @@ START_TEST(NtQueryObject)
...
+    HANDLE FileHandle;
+    FileHandle = CreateFileW(ExecutablePath,
+                             GENERIC_READ,
+                             FILE_SHARE_READ | FILE_SHARE_WRITE,
+                             NULL,
+                             OPEN_EXISTING,
+                             FILE_ATTRIBUTE_NORMAL,
+                             NULL);
+    ok(FileHandle != INVALID_HANDLE_VALUE,
+        "File '%S': Opening failed\n", ExecutablePath);
```
Comment (HBelusca) [inline modules/rostests/apitests/ntdll/NtQueryObject.c]:
```suggestion
    ok(FileHandle != INVALID_HANDLE_VALUE,
       "File '%S': Opening failed\n", ExecutablePath);
```

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+    ok_err(0xdeadbeef);
+
+    /* Clean up */
+    CloseHandle(hFile);
+    Status = NetShareDel(L"", ShareName, 0);
+}
+
+START_TEST(GetFinalPathNameByHandle)
+{
+    HMODULE hmodKerne32 = GetModuleHandleW(L"kernel32.dll");
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: "hmodKerne32" ---> "hmodKernel32"

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+    WCHAR ExpectedString[MAX_PATH];
+    PWSTR FileName;
+    DWORD Result;
+    SIZE_T ExpectedStringLength;
+    DWORD dwError = 0;
+    HANDLE hFile;
+    NET_API_STATUS Status;
+
+    /* Get the full path name of the current executable */
+    Result = GetModuleFileNameW(NULL, PathBuffer, MAX_PATH);
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]:
```suggestion
    Result = GetModuleFileNameW(NULL, PathBuffer, ARRAYSIZE(PathBuffer));
```

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+    shareInfo.shi2_max_uses = -1; // Unlimited connections
+    shareInfo.shi2_current_uses = 0; // 0 for new share
+    shareInfo.shi2_path = PathBuffer;
+    shareInfo.shi2_passwd = NULL; // No password
+
+    // Call NetShareAdd to create the share
+    Status = NetShareAdd(L"", // Empty string for local machine
+                         2,             // Level 2 for SHARE_INFO_2
+                         (LPBYTE)&shareInfo, // Share info structure
+                         &dwError);       // Error code if applicable
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c:253]: not sure we need the comments for each of the parameters ^^

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_NT | FILE_NAME_OPENED);
+    ok_eq_ulong(Result, ExpectedStringLength);
+    ok_eq_wstr(Buffer, ExpectedString);
+    ok_err(0xdeadbeef);
+
+    /* Expected string for VOLUME_NAME_NONE:
+       L"\\localhost\\TestShare\\kernel32_apitest.exe"*/
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]:
```suggestion
       L"\\localhost\\TestShare\\kernel32_apitest.exe" */
```

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_GUID | FILE_NAME_OPENED);
+    ok_eq_ulong(Result, 0ul);
+    ok_eq_wchar(Buffer[0], 0xCCCC);
+    ok_err(ERROR_PATH_NOT_FOUND);
+
+    /* Expected string for VOLUME_NAME_NT (2):
+       L"\\Device\\Mup\\localhost\\TestShare\\kernel32_apitest.exe"*/
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]:
```suggestion
       L"\\Device\\Mup\\localhost\\TestShare\\kernel32_apitest.exe" */
```

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_NT | FILE_NAME_OPENED);
+    ok_eq_ulong(Result, ExpectedStringLength);
+    ok_int(wcsncmp(Buffer, L"\\Device\\HarddiskVolume", 22), 0);
+    ok_int(wcscmp(Buffer + 24, VolumeRelativePath), 0);
+    ok_err(0xdeadbeef);
+
+    /* Expected string for VOLUME_NAME_NONE:
+       L"\\ReactOS\\bin\\kernel32_apitest.exe"*/
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]:
```suggestion
       L"\\ReactOS\\bin\\kernel32_apitest.exe" */
```

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 15]
```diff
@@ -0,0 +1,375 @@
...
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_GUID | FILE_NAME_OPENED);
+    ok_eq_ulong(Result, ExpectedStringLength);
+    ok_eq_wstr(Buffer, ExpectedString);
+    ok_err(0);
+
+    /* Expected string for VOLUME_NAME_NT (2):
+       L"\\Device\\HarddiskVolume1\\ReactOS\\bin\\kernel32_apitest.exe"*/
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]:
```suggestion
       L"\\Device\\HarddiskVolume1\\ReactOS\\bin\\kernel32_apitest.exe" */
```

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 16]
```diff
@@ -0,0 +1,375 @@
...
+    SIZE_T ExpectedStringLength;
+    HANDLE hFile;
+    BOOL Success;
+
+    /* Get the full path name of the current executable */
+    Result = GetModuleFileNameW(NULL, FilePath, MAX_PATH);
+    ok(Result != 0, "GetModuleFileNameW failed: %ld\n", GetLastError());
+    if (Result == 0)
+    {
+        skip("GetModuleFileNameW failed: %ld\n", GetLastError());
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: nit: since in case of Result == 0, we've got the previous error showing up, I don't think you need to repeat it again there; maybe just say "Skipping Test_File tests".

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 17]
```diff
@@ -0,0 +1,375 @@
...
+    WCHAR VolumeBuffer[4];
+    PWSTR VolumeRelativePath;
+    WCHAR ExpectedString[MAX_PATH];
+    DWORD Result;
+    SIZE_T ExpectedStringLength;
+    HANDLE hFile;
+    BOOL Success;
+
+    /* Get the full path name of the current executable */
+    Result = GetModuleFileNameW(NULL, FilePath, MAX_PATH);
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]:
```suggestion
    Result = GetModuleFileNameW(NULL, FilePath, ARRAYSIZE(FilePath));
```

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+
+    /* Query the GUID based volume name */
+    memcpy(VolumeBuffer, FilePath, 3 * sizeof(WCHAR));
+    VolumeBuffer[3] = UNICODE_NULL;
+    Success = GetVolumeNameForVolumeMountPointW(VolumeBuffer,
+                                                ExpectedString,
+                                                ARRAYSIZE(ExpectedString));
+    if (!Success)
+    {
+        trace("GetVolumeNameForVolumeMountPointW failed: %ld\n", GetLastError());
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: trace? not skip()?

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+        skip("Failed to create share. Error code: %ld\n", Status);
+        if (Status == ERROR_ACCESS_DENIED)
+        {
+            wprintf(L"Error: Access denied. Ensure you have administrative privileges.\n");
+        }
+
+        return;
+    }
+
+    swprintf(RemotePathBuffer, L"\\\\localhost\\%s\\%s", ShareName, FileName);
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c:265]: RemotePathBuffer being MAX_PATH will be large enough? What about StringCchPrintfW?

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c [hunk 20]
```diff
@@ -0,0 +1,289 @@
...
+                                   sizeof(Buffer),
+                                   Buffer,
+                                   sizeof(Buffer));
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("NtDeviceIoControlFile failed: %lx\n", Status);
+        goto Exit;
+    }
+
+    if ((IoStatusBlock.Information < sizeof(MOUNTMGR_VOLUME_PATHS)) ||
```
Comment (HBelusca) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]:
```suggestion
    if ((IoStatusBlock.Information < sizeof(*VolumePaths)) ||
```

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 1 above)
Comment (HBelusca) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: In case the NtClose call that follows could modify the last error, I may suggest you do a `BaseSetLastNTError(Status);` instead just before the `return Status;` below? Would this pass the tests as well?

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c [hunk 21]
```diff
@@ -0,0 +1,289 @@
...
+        return 0;
+    }
+
+    /* Make sure the object name is a valid device Path */
+    if ((NameInfo->Name.Buffer == NULL) ||
+        !RtlPrefixUnicodeString(&DevicePrefix, &NameInfo->Name, TRUE))
+    {
+        /* The object name is not valid */
+        DPRINT1("Invalid object name: %wZ\n", &NameInfo->Name);
+        SetLastError(ERROR_INVALID_HANDLE);
```
Comment (HBelusca) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: Invalid handle? `ERROR_INVALID_NAME` would be more suitable there. The new test-cases appear to only test the last-error being `ERROR_INVALID_HANDLE` whenever the `hFile` parameter has been set to either NULL or `INVALID_FILE_HANDLE` (or another invalid handle).
If we are there, the NtQueryObject call succeeded so the hFile handle was valid.

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c [hunk 22]
```diff
@@ -0,0 +1,289 @@
...
+        /* Return the length of the name */
+        return FinalLength;
+    }
+
+    /* For everything else we need to split the full path */
+    PathSplit = wcschr(&NameInfo->Name.Buffer[sizeof("\\Device\\") - 1], L'\\');
+    if (PathSplit == NULL)
+    {
+        DPRINT1("Invalid object name: %wZ\n", &NameInfo->Name);
+        SetLastError(ERROR_INVALID_HANDLE);
```
Comment (HBelusca) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c:226]: See comment above.

**Comment by DarkFire01** [comment]
image
Allows a factorio space age to load into the game now instead of crashing with a windows plugin error.
(A windows 7 SP1+ 64-bit game game(

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 3 above)
Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c:38]: This function is only used from below and that validates everything before calling it.

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 1 above)
Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: I changed to `ERROR_SUCCESS`. It is needed here because it is set in the `VOLUME_NAME_DOS` path, but not for network shares (which are skipping this path). I don't see how `BaseSetLastNTError(Status)` would improve anything.

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 19 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c:265]: The names are all known. If you believe that "\\\\localhost\\TestShare\\kernel32_apitest.exe" is larger than MAX_PATH, you need to up your estimation skills.

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 4 above)
Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: Fixed

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 2 above)
Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: GetVolumeNameForVolumeMountPointW does modify last error, but it sets an actual error value for some reason.

Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: Fixed

File: modules/rostests/apitests/ntdll/NtQueryObject.c (hunk 6 above)
Comment (tkreuzer) [reply modules/rostests/apitests/ntdll/NtQueryObject.c]: Fixed

File: modules/rostests/apitests/ntdll/NtQueryObject.c (hunk 7 above)
Comment (tkreuzer) [reply modules/rostests/apitests/ntdll/NtQueryObject.c]: Fixed

File: modules/rostests/apitests/ntdll/NtQueryObject.c (hunk 8 above)
Comment (tkreuzer) [reply modules/rostests/apitests/ntdll/NtQueryObject.c]: Fixed

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 9 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: Fixed

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 10 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: Done

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 17 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: Done

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 20 above)
Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: Done

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 21 above)
Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c]: I added more tests, including the Null device. To handle this the code now calls `NtQueryInformationFile`, which validates the file handle, which takes care of the error code. This part has been removed, because it isn't required anymore.

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 22 above)
Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c:226]: This path is an internal error and can probably never happen.

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 18 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: Fixed

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 16 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: Removed

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 15 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: Fixed

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 14 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: Fixed

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 13 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: Fixed

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 12 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c]: Fixed

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 11 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c:253]: not sure if I care ^^

**Review by DarkFire01 — APPROVED** [review]

File: modules/rostests/apitests/ntdll/NtQueryObject.c [hunk 23]
```diff
@@ -99,9 +99,42 @@ START_TEST(NtQueryObject)
...
     }
+
     /* Verify that ObjectName->Length doesn't count extra NUL-terminators */
-    {
     SIZE_T strLen = wcslen(ObjectName->Buffer) * sizeof(WCHAR);
     ok_eq_size(strLen, (SIZE_T)ObjectName->Length);
+
+    /* Get the full path name of the current executable */
+    WCHAR ExecutablePath[MAX_PATH];
+    GetModuleFileNameW(NULL, ExecutablePath, MAX_PATH);
```
Comment (HBelusca) [inline modules/rostests/apitests/ntdll/NtQueryObject.c:109]:
```suggestion
    GetModuleFileNameW(NULL, ExecutablePath, _countof(ExecutablePath));
```
or with ARRAYSIZE

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c [hunk 24]
```diff
@@ -0,0 +1,298 @@
...
+    ULONG VolumeType;
+    NTSTATUS Status;
+    ULONG FinalLength;
+    BOOL Success;
+
+    /* Extract the requested volume type */
+    VolumeType = dwFlags & (VOLUME_NAME_GUID | VOLUME_NAME_NONE | VOLUME_NAME_NT);
+
+    /* Check if more than one volume type flag is set */
+    if (VolumeType & (VolumeType- 1))
```
Comment (HBelusca) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c:150]:
```suggestion
    if (VolumeType & (VolumeType - 1))
```

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c [hunk 25]
```diff
@@ -0,0 +1,298 @@
...
+        lpszFilePath[FinalLength] = UNICODE_NULL;
+
+        /* Return the length of the name */
+        return FinalLength;
+    }
+    else if (VolumeType == VOLUME_NAME_GUID)
+    {
+        /* Query the GUID volume name */
+        Success = GetVolumeNameForVolumeMountPointW(VolumeBuffer,
+                                                    VolumeBuffer,
```
Comment (HBelusca) [inline dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c:272]: Is it valid to invoke the function on the same buffer for both the input and the output parameter? It may appear to work but usually such a thing could be undefined behaviour (dependent on the exact implementation of GetVolumeNameForVolumeMountPointW).

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 24 above)
Comment (HBelusca) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c:150]: The function doesn't verify if other unknown flags are passed?
For example in the apitest, this line:
```c
pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_GUID | FILE_NAME_OPENED);
```
which appears to succeed: what happens if you add `0xFFFFFFF0` to the flags?

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c [hunk 26]
```diff
@@ -0,0 +1,464 @@
...
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_GUID);
+    ok_eq_ulong(Result, ExpectedStringLength);
+    ok_eq_wstr(Buffer, ExpectedString);
+    ok_err(0);
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_GUID | FILE_NAME_OPENED);
```
Comment (HBelusca) [inline modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c:152]: Your implementation doesn't seem to do anything with either `FILE_NAME_NORMALIZED` or `FILE_NAME_OPENED`.
Any reason why so (missing feature in the kernel or in kernel32/kernelbase, ...)?

**Review by HBelusca — COMMENTED** [review]
Test results with the current implementation:
- Test KVM: https://reactos.org/testman/compare.php?ids=102496,102502
  [67 tests executed, 44 failures, 1 skipped](https://reactos.org/testman/detail.php?id=82188313&prev=0)
- Test KVM_64: https://reactos.org/testman/compare.php?ids=102497,102503
  [67 tests executed, 0 failures, 1 skipped](https://reactos.org/testman/detail.php?id=82189589&prev=0)

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 25 above)
Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c:272]: Yes, it's valid.

File: dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c (hunk 24 above)
Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c:150]: Nothing, they are ignored.

File: modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c (hunk 26 above)
Comment (tkreuzer) [reply modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c:152]: Those don't do anything that was visible. Feel free to write a test that shows that they do something.
